### PR TITLE
Sync Server

### PR DESCRIPTION
**Pull request by kalisp**
Requires:
pype-config:pypeclub/pype-config#82
pype-setup:pypeclub/pype-setup#73

Basic implementation of synchronization server to/from GDrive.

It expects this format in representations in DB:

`
"sites" : {

                "studio" : {

                    "created_dt" : ISODate("2020-05-22T08:05:44.000Z")

                },
                "gdrive" : {

                    "id" : ObjectId("5eeb25e411e06a16209ab78f"),

                    "created_dt" : ISODate("2020-07-16T17:54:35.833Z")

                }
            }`

'studio', 'gdrive' denotes destination. 

Studio should be local version and should always be present (at least if user published files).
Process looks for presence of "created_dt". 
If not found or empty, it tries to sync to that destination from source (site that has 'created_dt' filled in. In this approach synching to and from GDrive is implemented already. If multiple sites have that file, first is picked up.)
Destination follows same folder structure as source.

Re-upload|re-download could be driven just by manually purging 'created_dt' for selected files.

It is expected that each provider (GDrive, S3) can have different limits (how many queries to their API), currently implemented some limits for GDrive. (It uploads files in batches that shouldn't hit limit - which will probably not work in case of sharing single user with multiple Trays.)
(Not implemented bandwidth throttling yet though.)

Syncing is using asyncio and shouldn't be blocking main thread. Sync server lives in Tray and checks database for files that should be uploaded periodically.

Logging info is stored in each representation in DB (in case of failure). Automatic retries are implemented with configurable limit.

No other front end is implemented yet.

Known issues: 
Authentication is not finished, it depends on InstalledAppFlow (where some human intervention is needed via browser).
(See https://developers.google.com/drive/api/v3/quickstart/python for more information. `credentials.json` from Cloud platform is needed). Possibility of service account is being investigated.

### CODE REVIEW

**Comment by kalisp** [comment]
**Notes:**

- Each user should be as separate site (eg `john_local_as3345f`, `ben_local_fds34g`)
- There is a difference between my and `to_synch_to` sites (eg. GDrive, S3)
- Tray is interested only in itself site (eg. `john_local_as3345f`) and ALL `to_sync_to` sites
- During publish representation is created with `my` site and filled `created_dt`. + skeleton for `to_sync_to` site, without `created_dt`
- Sync server in next check notices that there is `to_sync_to` site without `created_dt` (but with `my` site present) and it will try to synchronize it (upload).
- Loader should highlight that file is on `to_sync_to` location, I myself can trigger "Download to my local" (that will update representation, it will add my site with empty `created_dt` >> sync will pick it up in next loop.)
- In the future some admin could force sync to specific users (by creating site(s) skeletons for all appropriate users) (via Ftrack action, maybe via Tray and admin permissions)

**Possible issues:**

- Performance (thousands of representations + tens of sites) - will be tested on dummy
- hitting on limits if all users (sites) will be sharing same account (in that case some global queue would need to be introduced - in that case user wont be checking DB, but queue for tasks for download)
- Performance of uploads/downloads, bandwidth

File: pype/modules/sync_server/sync_server.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+                                    task_files_to_process.append(task)
+                                    # store info for exception handling
+                                    files_processed_info.append((file,
+                                                                 sync,
+                                                                 provider))
+                                    processed_file_path.add(file_path)
+                                if status == SyncStatus.DO_DOWNLOAD:
+                                    limit -= 1
+                                    task = asyncio.create_task(
+                                                   self.module.download
```
[the remark this answers is not in the document]
Comment (kalisp) [reply pype/modules/sync_server/sync_server.py:588]: Ohhh, how I hate you for these...

**Comment by kalisp** [comment]
Structure of 'sites' changed to:
`  "sites" : [
                {
                    "name": "john_local_XD4345",
                    "created_dt" : ISODate("2020-05-22T08:05:44.000Z")
                },
                {
                    "id" : ObjectId("5eeb25e411e06a16209ab78f"),
                    "name": "gdrive",
                    "created_dt" : ISODate("2020-05-55T08:54:35.833Z")
                ]
            }`

for better performance.

**Comment by kalisp** [comment]
Some notes about use cases: (based on 21/9/2020 Weekly)

> user should be able to:
>>select 'primary' sync site (different sync site when she is working in studio|home)
>>disable local site (publish only to 'studio' site - mapped studio drive)
> user can publish to 'studio' (GDrive mapped), but close laptop immediately, eg. no sync done, file still only locally, maybe add 'verified' tag for peers to verify that file was actually uploaded(synced)

Some work needs to be done on 'integrate_new'  where in case of disablement of local site, implicit sync to remote should be triggered.

**Comment by kalisp** [comment]
Added active_site, remote_site to presets. These should be used as a default, exposed appropriate properties to be able to change it from GUI.

Adding couple of pictures how to configure Google Drive service account:
**Go to** https://console.developers.google.com/ and create project
sync_server_create_project

**For project** create service account
sync_server_create_service_account

sync_server_create_service_account2

**Fill only** first section, no need for rest, remember created email
sync_server_create_service_account3

**Create key** in JSON format, store it anywhere Pype has access and put its path to `pype-config/presets/sync_server/gdrive.json`
sync_server_create_service_account4


**Create folder** with name matching to project on Ftrack, share it with email of service account
sync_server_gdrive_share

Publish any representation, sites for 'studio' and 'gdrive' should be created automatically (based on  `pype-config/presets/sync_server/config.json`)

All possible errors should be captured under appropriate site,  any failed file will be retried 3x times by default.

**Comment by kalisp** [comment]
TODO:
Implement "root" field in `gdrive.json` to handle multiroot projects. Configuration should follow existing multiroot config.
{root} placeholder should be replaced by value from `gdrive.json`.

**Review by mkolar — COMMENTED** [review]
I've done quite a lot of testing on this yesterday and today and it seems quite solid. Needs implementation for root and multiroot as mentioned by @kalisp, but other than that it is ready for alpha stage testing.

**Comment by kalisp** [comment]
Implemented multiroot and multisite.

Multiroot:
Different folders on same GDrive
Config:
```
"root": {
            "root_one": "/My Drive/work_folder",
            "root_tow": "/My Drive/publish_folder"
      }
```


Multisite:
Different accounts for GDrive implementation
Config:
```
{
  "gdrive1": {
    "credentials_url": "/mysecretfolder/credentials1.json",
    "root": "/My Drive"
  },
  "gdrive2": {
    "credentials_url": "/mysecretfolder/credentials2.json",
    "root": "/My Drive"
  }
}
```

**Comment by mkolar** [comment]
After the last change I can't get it to sync. It might be due to wrong config. Just to clarify. If I want to use a drive folder name `myPypeRoot` as a single root I need to add.
```
"gdrive": {
    "credentials_url": "/mysecretfolder/credentials1.json",
    "root": "/myPypeRoot"
  },
```
The publishing is working correctly and I get skeleton site data in the db, but it seems that the sync server doesn't quite pick it up, to even process. This is what my sites look like on a representation file.
image

**Comment by mkolar** [comment]
Another strange thing is happening. When I run pype with this branch it immediately slows down my machine quite a bit. It's even visible in less sensitive and slow touchpad on the laptop. Feel like we have an ugly leak somewhere.

**Comment by kalisp** [comment]
Weird, couldn't reproduce on my end.
I kept Pype running for a while and didn't notice memory footprint creeping up.

Please double check Pype logs if SyncServer is actually running, check for this "No presets or active providers. Synchronization not possible.".

Btw. try "root": "/My Drive/myPypeRoot". I am not sure if "My Drive" is implicit and it could/should be omitted, or if different GDrives could have different root folder names. I went for explicit vs implicit.

For you content of gdrive.json should be 
```
{
  "gdrive": {
    "credentials_url": "/mysecretfolder/credentials1.json",
    "root": "/My Drive/myPypeRoot"
  }
}
```

**Comment by mkolar** [comment]
I get this in the log but only at the start and then nothing further from the server.
image

Really strange that it was working for me before without issues and now nothing.

**Comment by mkolar** [comment]
After some investigation we found out that the problem was caused by gdrive bin system. Some older published files have been deleted with their root folder and the sync process couldn't make sense of the fact, that the IDs were still existing, but they were technically orphaned. Deleting them from the bin solve the problem, but Petr is working on more robust checks here. 

I've also tested multiroot setup and it seem to work very well. The one thing I didn't manage is to use this with google `shared drives`. It only finds the correct roots when using `my drive` style roots. We'll have to look into that, considering shared drives are somewhat more appropriate for our usage.

**Comment by kalisp** [comment]
Implemented support for shared disks.
Root url for them should be "/SHARED_DISK_NAME/..." 

'/My Drive' is still there for regular shared folders

**Comment by mkolar** [comment]
Works pretty well. I'm happy to merge as alpha feature, so we can start some bigger testing. Maybe we could push it to 2x/develop as well, so get more testing under way.

**Pull request merged**